### ISSUE TITLE
GreedyTDTInfer skips an extra encoder frame when the symbol budget is exhausted

### ISSUE THREAD

**Issue by udsy19** [NVIDIA-NeMo/NeMo#16095]
**Describe the bug**

`GreedyTDTInfer` (TDT models decoded with `strategy: greedy`) advances the encoder
time index one frame too far whenever the per-frame symbol budget is exhausted.
With `max_symbols_per_step=1` the budget is exhausted on *every* frame, so the
decoder skips `duration + 1` frames each step: tokens are dropped and every
surviving timestamp is late.

`nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:2726-2747`
(`GreedyTDTInfer._greedy_decode`):

```python
                # Increment token counter.
                symbols_added += 1
                time_idx += skip          # TDT: advance by the predicted duration
                need_loop = skip == 0

            # this rarely happens, but we manually increment the `skip` number
            # if blank is emitted and duration=0 is predicted. This prevents possible
            # infinite loops.
            if skip == 0:
                skip = 1
            ...
            if symbols_added == self.max_symbols:
                time_idx += 1             # fires even when time has already advanced
```

The inner loop exits either because a non-zero duration was predicted (`need_loop`
is `False`, and `time_idx` has already been advanced by that duration) or because
the symbol budget ran out. The forced `time_idx += 1` is only needed in the second
case *when the last prediction had duration 0* — but as written it fires in the
first case too.

**Steps/Code to reproduce bug**

CPU-only, no downloads, no checkpoint. A constant joint makes the correct answer
computable by hand: the model always emits token 1 with duration 2, so with `T`
frames the tokens must land at `t = 0, 2, 4, ...`.

```python
import torch
import nemo.collections.asr.parts.submodules.rnnt_greedy_decoding as gd
from nemo.collections.asr.modules import RNNTDecoder, RNNTJoint

TOKENS = [" ", "a", "b", "c"]
V, BLANK, D = len(TOKENS), len(TOKENS), 4
DURATIONS = [0, 1, 2, 4]


class ConstJoint(RNNTJoint):
    def joint_after_projection(self, f, g):
        B = f.shape[0]
        T = f.shape[1] if f.dim() == 3 else 1
        U = g.shape[1] if g.dim() == 3 else 1
        out = torch.full((B, T, U, (V + 1) + len(DURATIONS)), -10.0)
        out[..., 1] = 5.0                    # always token 1
        out[..., (V + 1) + 2] = 5.0          # always duration DURATIONS[2] == 2
        return out


def run(cls, ms, T, **kw):
    torch.manual_seed(0)
    dec = RNNTDecoder({'pred_hidden': D, 'pred_rnn_layers': 1}, V)
    jnt = ConstJoint({'encoder_hidden': D, 'pred_hidden': D, 'joint_hidden': D, 'activation': 'relu'},
                     V, vocabulary=TOKENS, num_extra_outputs=len(DURATIONS))
    dec.eval(); jnt.eval()
    d = cls(dec, jnt, blank_index=BLANK, durations=DURATIONS,
            max_symbols_per_step=ms, include_duration=True, **kw)
    with torch.no_grad():
        h = d(encoder_output=torch.zeros(1, D, T),
              encoded_lengths=torch.tensor([T], dtype=torch.int32))[0][0]
    return list(map(int, h.timestamp)), list(map(int, h.token_duration))


T = 10
print("expected              :", list(range(0, T, 2)))
ts, dur = run(gd.GreedyTDTInfer, 1, T)
print("GreedyTDTInfer  ms=1  :", ts, "token_duration:", dur)
ts, dur = run(gd.GreedyBatchedTDTInfer, 1, T, use_cuda_graph_decoder=False)
print("GreedyBatchedTDT ms=1 :", ts, "token_duration:", dur)
```

Output:

```
expected              : [0, 2, 4, 6, 8]
GreedyTDTInfer  ms=1  : [0, 3, 6, 9] token_duration: [2, 2, 2, 2]
GreedyBatchedTDT ms=1 : [0, 2, 4, 6, 8] token_duration: [2, 2, 2, 2, 2]
```

`GreedyTDTInfer` emits 4 tokens where 5 are correct, and every timestamp after the
first is late. It still reports `token_duration = [2, 2, 2, 2]`, so the hypothesis
is internally inconsistent: `timestamp[k] + token_duration[k] != timestamp[k + 1]`.

`preserve_alignments=True` on the same input shows the same desync from the other
side — the alignment buffer is built from `skip` while the time index advances by
`skip + 1`:

```
shipped:  T=10  len(alignments)=8   timestamps=[0, 3, 6, 9]   non-empty slots=[0, 2, 4, 6]
```

Ten encoder frames produce only eight alignment slots, and the slots that hold
logits (`0, 2, 4, 6`) do not agree with the reported timestamps (`0, 3, 6, 9`).

**It is not limited to `max_symbols_per_step=1`.** The budget can also be exhausted
on a non-zero duration when an earlier step in the same frame predicted duration 0.
With a joint that alternates duration 0 and duration 2 on successive calls and
`max_symbols_per_step=2`, the correct timestamps are `[0, 0, 2, 2, 4, 4, 6, 6, 8, 8]`
and `GreedyTDTInfer` returns `[0, 0, 3, 3, 6, 6, 9, 9]`.

Over a sweep of random weights (3 seeds x 3 duration sets x 3 batch shapes x
`max_symbols_per_step` in {1, 2, 3, 5} = 468 utterance comparisons),
`GreedyTDTInfer` disagreed with `GreedyBatchedTDTInfer` on `(tokens, timestamps)`
for 135 utterances — 111/117 at `max_symbols_per_step=1`, and 9, 9 and 6 at 2, 3
and 5 respectively.

**Expected behavior**

`GreedyTDTInfer` should advance the time index by exactly the predicted duration,
and force a one-frame advance only when the predicted duration was 0 and the symbol
budget stopped the loop — which is what `GreedyBatchedTDTInfer` does.

Since the inner loop can only exit with `skip == 0` because the budget ran out, the
forced advance folds into the existing zero-duration branch and the trailing
conditional can be deleted:

```python
            if skip == 0:
                skip = 1
                time_idx += 1
            ...
            # if symbols_added == self.max_symbols:   <- delete
            #     time_idx += 1
```

With that change all 468 comparisons above agree, and the alignment buffer length
matches the encoder length again.

**Environment overview**

- Environment location: Bare-metal (macOS arm64), CPU only
- Method of NeMo install: from source, `main` @ `d0a07c056`

**Environment details**

- PyTorch 2.12.0
- Python 3.10.18
